README.md
# chef-cfsolr [![Build Status](https://secure.travis-ci.org/wharton/chef-cfsolr.png?branch=master)](http://travis-ci.org/wharton/chef-cfsolr)

## Description

Installs/Configures Apache Solr standalone for Adobe ColdFusion.

## Requirements

### Platforms

* CentOS 6.3
* RHEL 6.3
* Ubuntu 12.04

## Attributes

These attributes are under the `node['cfsolr']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
checksum | ColdFusion Solr download checksum | String | see attributes/default.rb
install_path | Solr installation path | String | /opt/coldfusionsolr
java_home | JVM for usage | String | `#{node['cfsolr']['install_path']}/jre`
java_max_heap | JVM maximum heap memory | String | 256m (CF9) or 512m (CF10)
solrport | Solr port | Fixnum | 8983 (CF9) or 8985 (CF10)
stopport | Solr stop port | Fixnum | 8079 (CF9) or 8077 (CF10)
url | ColdFusion Solr download URL | String | see attributes/default.rb
version | ColdFusion Solr version | String | 10.0.0

### Replication Attributes

These attributes are under the `node['cfsolr']['master']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
enabled | Node is master server | Boolean | true
hostname | Hostname of master server | String | `node['fqdn']`
port | Port of master server | Fixnum | `node['cfsolr']['solrport']`

These attributes are under the `node['cfsolr']['slave']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
pollInterval | Solr poll interval | String | 00:00:60

## Recipes

* `recipe[cfsolr]` - Installs/configures Apache Solr standalone for Adobe ColdFusion
* `recipe[cfsolr::replication]` - Simple replication configuration for ColdFusion Solr servers

## Usage

### Solr Standalone Server Installation

* Add `recipe[cfsolr]` to your node's run list

### Simple Solr Replication Server Setup

* On master nodes:
  * Add `recipe[cfsolr]` and `recipe[cfsolr::replication]` to run list
* On slave nodes:
  * Add `recipe[cfsolr]` and `recipe[cfsolr::replication]` to run list
  * Set `node['cfsolr']['master']['enable']` to false
  * Set `node['cfsolr']['master']['hostname']` to FQDN of master node
  * Adjust `node['cfsolr']['slave']` attributes as necessary

## Testing and Development

Here's how you can quickly get testing or developing against the cookbook thanks to [Vagrant](http://vagrantup.com/) and [Berkshelf](http://berkshelf.com/).

    gem install bundler --no-ri --no-rdoc
    git clone git://github.com/wharton/chef-cfsolr.git
    cd chef-cfsolr
    bundle install
    bundle exec vagrant up BOX # BOX being centos6 or ubuntu1204

The running ColdFusion Solr server is accessible from the host machine:

* Web UI: http://33.33.33.10:8983/solr/admin/cores?action=status

You can then SSH into the running VM using the `vagrant ssh` command.
The VM can easily be stopped and deleted with the `vagrant destroy`
command. Please see the official [Vagrant documentation](http://vagrantup.com/v1/docs/commands.html)
for a more in depth explanation of available commands.

## Contributing

Please use standard Github issues/pull requests and if possible, in combination with testing on the Vagrant boxes.

## License and Author

Author:: Nathan Mische (<nmische@wharton.upenn.edu>), Brian Flad (<bflad@wharton.upenn.edu>)

Copyright:: 2011-2013

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## v0.3.0

* Added `node['cfsolr']['master']['port']` attribute and configuration

## v0.2.0

* Corrected default ColdFusion 10 Solr port and JVM Xmx
* Added `solrport` and `stopport` attributes
* Detect init script name based on version
* Detect installation on file inside install_path (init script) rather than just install_path
* Correctly handle init script changes per version
* Renamed init script template to cfsolr.init.erb

## v0.1.0

* Support for ColdFusion Solr 9.0.0, 9.0.1, 10.0.0
* Support for CentOS/RHEL 6.3
* Added simple replication recipe

## v0.0.1

* Initial release


recipes/replication.rb
#
# Cookbook Name:: cfsolr
# Recipe:: master
#
# Copyright 2013, Brian Flad
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Solr configuration
template "#{node['cfsolr']['install_path']}/multicore/template/conf/solrconfig.xml" do
  source "solrconfig.xml.erb"
  mode "0644"
  notifies :restart, "service[cfsolr]", :delayed
end

# Solr configuration
template "#{node['cfsolr']['install_path']}/multicore/template/conf/solrcore.properties" do
  source "solrcore.properties.erb"
  mode "0644"
  notifies :restart, "service[cfsolr]", :delayed
end


attributes/default.rb

default["cfsolr"]["version"] = "10.0.0"

case Chef::Version.new(node["cfsolr"]["version"]).major
when 10
  #default["cfsolr"]["url"] = "http://download.macromedia.com/pub/coldfusion/updates/10/SolrInstallers/ColdFusion_10_Jetty_Solr_linux.bin"
  #default["cfsolr"]["checksum"] = "4137ee93041139371fa73ed7ad48c781ea80cfed6742f296f1c1c4a2015b5a0d"
  default["cfsolr"]["url"] = "http://download.macromedia.com/pub/coldfusion/updates/10/SolrInstallers/ColdFusion_10_Jetty_Solr_linux64.bin"
  default["cfsolr"]["checksum"] = "c3f33aec960e44fd1ec87df3e72f19addd43aaa732e03ae73ab5293eac599f9f"
  default["cfsolr"]["java_max_heap"] = "512m"
  default["cfsolr"]["solrport"] = 8985
  default["cfsolr"]["stopport"] = 8077
when 9
  case Chef::Version.new(node["cfsolr"]["version"]).patch
  when 1
    default["cfsolr"]["url"] = "http://download.macromedia.com/pub/coldfusion/updates/901/ColdFusion_901_Solr_WWEJ_linux64.bin"
    default["cfsolr"]["checksum"] = "807b8fbdc0d7dfe841127fc8f7affd178d715f1a39e12ba99a7442c01c7ee9a7"
  when 0
    default["cfsolr"]["url"] = "http://download.macromedia.com/pub/coldfusion/9/solr_installer/ColdFusion_9_Solr_linux64.bin"
    default["cfsolr"]["checksum"] = "680aa4d70c645412de65e0370d78e93b23cfbffd360046b3f2f71363207342d6"
  end
  default["cfsolr"]["java_max_heap"] = "256m"
  default["cfsolr"]["solrport"] = 8983
  default["cfsolr"]["stopport"] = 8079
end

# CF Solr installation folder
default["cfsolr"]["install_path"] = "/opt/coldfusionsolr"

# JVM
default["cfsolr"]["java_home"] = "#{node['cfsolr']['install_path']}/jre"

# Replication
default["cfsolr"]["master"]["enable"] = true
default["cfsolr"]["master"]["hostname"] = node["fqdn"]
default["cfsolr"]["master"]["port"] = node["cfsolr"]["solrport"]
default["cfsolr"]["slave"]["pollInterval"] = "00:00:60"
